### ISSUE TITLE
No eligible torsions found during geometry engine proposal

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#160]
It's back!

```
--------------------------------------------------------------------------------
Expanded Ensemble sampler iteration       18
................................................................................
MCMC sampler iteration 18
Taking 20000 steps of Langevin dynamics...
potential  = -6652.846 kT
kinetic    = 6587.012 kT
force norm = 36.346 kT/A/dof
Final energy is    -6679.401 kT
................................................................................
Updating chemical state with ncmc-geometry-ncmc scheme...
Proposing new topology...
Proposed mutation:
[('37', 'TYR')]
Proposed transformation: TRP-GLU-MET-GLU-ARG-THR-ASP-ILE-THR-MET-LYS-HIS-LYS-LEU-GLY-GLY-GLY-GLN-TYR-GLY-GLU-VAL-TYR-GLU-GLY-VAL-TRP-LYS-LYS-TYR-SER-LEU-THR-VAL-ALA-VAL-LYS-THR-LEU-LYS-GLU-ASP-THR-MET-GLU-VAL-GLU-GLU-PHE-LEU-LYS-GLU-ALA-ALA-VAL-MET-LYS-GLU-ILE-LYS-HIS-PRO-ASN-LEU-VAL-GLN-LEU-LEU-GLY-VAL-CYS-THR-ARG-GLU-PRO-PRO-PHE-TYR-ILE-ILE-THR-GLU-PHE-MET-THR-TYR-GLY-ASN-LEU-LEU-ASP-TYR-LEU-ARG-GLU-CYS-ASN-ARG-GLN-GLU-VAL-ASN-ALA-VAL-VAL-LEU-LEU-TYR-MET-ALA-THR-GLN-ILE-SER-SER-ALA-MET-GLU-TYR-LEU-GLU-LYS-LYS-ASN-PHE-ILE-HIS-ARG-ASP-LEU-ALA-ALA-ARG-ASN-CYS-LEU-VAL-GLY-GLU-ASN-HIS-LEU-VAL-LYS-VAL-ALA-ASP-PHE-GLY-LEU-SER-ARG-LEU-MET-THR-GLY-ASP-THR-TYR-THR-ALA-HIS-ALA-GLY-ALA-LYS-PHE-PRO-ILE-LYS-TRP-THR-ALA-PRO-GLU-SER-LEU-ALA-TYR-ASN-LYS-PHE-SER-ILE-LYS-SER-ASP-VAL-TRP-ALA-PHE-GLY-VAL-LEU-LEU-TRP-GLU-ILE-ALA-THR-TYR-GLY-MET-SER-PRO-TYR-PRO-GLY-ILE-ASP-LEU-SER-GLN-VAL-TYR-GLU-LEU-LEU-GLU-LYS-ASP-TYR-ARG-MET-GLU-ARG-PRO-GLU-GLY-CYS-PRO-GLU-LYS-VAL-TYR-GLU-LEU-MET-ARG-ALA-CYS-TRP-GLN-TRP-ASN-PRO-SER-ASP-ARG-PRO-SER-PHE-ALA-GLU-ILE-HIS-GLN-ALA-PHE-GLU-THR-MET-PHE-GLN-MOL => TRP-GLU-MET-GLU-ARG-THR-ASP-ILE-THR-MET-LYS-HIS-LYS-LEU-GLY-GLY-GLY-GLN-TYR-GLY-GLU-VAL-TYR-GLU-GLY-VAL-TRP-LYS-LYS-TYR-SER-LEU-THR-VAL-ALA-VAL-TYR-THR-LEU-LYS-GLU-ASP-THR-MET-GLU-VAL-GLU-GLU-PHE-LEU-LYS-GLU-ALA-ALA-VAL-MET-LYS-GLU-ILE-LYS-HIS-PRO-ASN-LEU-VAL-GLN-LEU-LEU-GLY-VAL-CYS-THR-ARG-GLU-PRO-PRO-PHE-TYR-ILE-ILE-THR-GLU-PHE-MET-THR-TYR-GLY-ASN-LEU-LEU-ASP-TYR-LEU-ARG-GLU-CYS-ASN-ARG-GLN-GLU-VAL-ASN-ALA-VAL-VAL-LEU-LEU-TYR-MET-ALA-THR-GLN-ILE-SER-SER-ALA-MET-GLU-TYR-LEU-GLU-LYS-LYS-ASN-PHE-ILE-HIS-ARG-ASP-LEU-ALA-ALA-ARG-ASN-CYS-LEU-VAL-GLY-GLU-ASN-HIS-LEU-VAL-LYS-VAL-ALA-ASP-PHE-GLY-LEU-SER-ARG-LEU-MET-THR-GLY-ASP-THR-TYR-THR-ALA-HIS-ALA-GLY-ALA-LYS-PHE-PRO-ILE-LYS-TRP-THR-ALA-PRO-GLU-SER-LEU-ALA-TYR-ASN-LYS-PHE-SER-ILE-LYS-SER-ASP-VAL-TRP-ALA-PHE-GLY-VAL-LEU-LEU-TRP-GLU-ILE-ALA-THR-TYR-GLY-MET-SER-PRO-TYR-PRO-GLY-ILE-ASP-LEU-SER-GLN-VAL-TYR-GLU-LEU-LEU-GLU-LYS-ASP-TYR-ARG-MET-GLU-ARG-PRO-GLU-GLY-CYS-PRO-GLU-LYS-VAL-TYR-GLU-LEU-MET-ARG-ALA-CYS-TRP-GLN-TRP-ASN-PRO-SER-ASP-ARG-PRO-SER-PHE-ALA-GLU-ILE-HIS-GLN-ALA-PHE-GLU-THR-MET-PHE-GLN-MOL
Generating new system...
Performing NCMC annihilation
alchemical atoms:
[590, 591, 592, 597, 598, 600, 602, 603, 604, 605]
LOG: Creating alchemically modified system...
LOG: particle 107 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 152 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 311 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 370 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 497 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 508 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 538 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 616 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 698 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1166 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1277 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1326 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1392 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1416 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1508 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1778 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1816 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1866 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1877 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 1940 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2452 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2523 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2556 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2580 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2591 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2780 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2833 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2883 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 2950 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3002 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3224 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3248 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3283 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3318 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3400 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3454 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3577 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3792 has Lennard-Jones sigma = 0 (charge=0.3992 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 3992 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 4053 has Lennard-Jones sigma = 0 (charge=0.4275 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: particle 4207 has Lennard-Jones sigma = 0 (charge=0.4102 e, sigma=0.0 nm, epsilon=0.0 kJ/mol); setting sigma=1A
LOG: Elapsed time 3.443 s.
LOG: Creating alchemically modified intermediate...
LOG: Elapsed time 2.504 s.
-16660.7094947 kJ/mol
-16660.7114511 kJ/mol
initial potential before 'delete' : -6679.402094 kT
initial potential components:   [('HarmonicBondForce', Quantity(value=6230.1534126943825, unit=kilojoule/mole)), ('HarmonicAngleForce', Quantity(value=9613.641207171831, unit=kilojoule/mole)), ('PeriodicTorsionForce', Quantity(value=13081.363981890625, unit=kilojoule/mole)), ('NonbondedForce', Quantity(value=-44962.4543219814, unit=kilojoule/mole)), ('CustomNonbondedForce', Quantity(value=-19.355205812618735, unit=kilojoule/mole)), ('CustomNonbondedForce', Quantity(value=-633.5521443474566, unit=kilojoule/mole)), ('CustomBondForce', Quantity(value=29.474847447431813, unit=kilojoule/mole)), ('CMMotionRemover', Quantity(value=0.0, unit=kilojoule/mole))]
final potential before 'delete' : -6415.283912 kT
final potential components: [('HarmonicBondForce', Quantity(value=6408.579056505032, unit=kilojoule/mole)), ('HarmonicAngleForce', Quantity(value=9535.057019751663, unit=kilojoule/mole)), ('PeriodicTorsionForce', Quantity(value=13152.64216762883, unit=kilojoule/mole)), ('NonbondedForce', Quantity(value=-45098.2006038415, unit=kilojoule/mole)), ('CustomNonbondedForce', Quantity(value=-16.51571914476393, unit=kilojoule/mole)), ('CustomNonbondedForce', Quantity(value=-607.5670574321343, unit=kilojoule/mole)), ('CustomBondForce', Quantity(value=27.97660035996908, unit=kilojoule/mole)), ('CMMotionRemover', Quantity(value=0.0, unit=kilojoule/mole))]

LOG: NCMC logP     -247.2 | initial_total_energy     -195.4 kT | final_total_energy     +421.3 kT.
Geometry engine proposal...
Traceback (most recent call last):
  File "perses/tests/testsystems.py", line 1007, in <module>
    run_abl_imatinib()
  File "perses/tests/testsystems.py", line 974, in run_abl_imatinib
    testsystem.exen_samplers[environment].run(niterations=100)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 804, in run
    self.update()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 784, in update
    self.update_state()
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 722, in update_state
    geometry_new_positions, geometry_logp_propose  = self.geometry_engine.propose(topology_proposal, geometry_old_positions, self.sampler.thermodynamic_state.beta)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 96, in propose
    logp_proposal, new_positions = self._logp_propose(top_proposal, current_positions, beta, direction='forward')
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 157, in _logp_propose
    atom_proposal_order, logp_choice = proposal_order_tool.determine_proposal_order(direction='forward')
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 871, in determine_proposal_order
    chosen_torsion, logp_choice = self._choose_torsion(atoms_with_positions, atom)
  File "/cbio/jclab/home/chodera/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/geometry.py", line 919, in _choose_torsion
    raise Exception("No eligible torsions found for placing atom %s." % str(atom_for_proposal))
Exception: No eligible torsions found for placing atom <Atom CE2 [603]; In TYR 36>.
```

**Comment by pgrinaway** [comment]
I realized last night that we're not being careful enough with how we propose the next state. For instance, in the small molecule case, while I have `p(molecule) \propto n_atoms_in_common`, this occasionally means that a molecule will be selected which has atoms in common, but no torsions connecting a group of new atoms, so the `GeometryEngine` fails. I assume the case is the same here, and imagine the resolution to involve two things:

* Make presence of torsions a requirement for proposing a transition (shouldn't be too hard, since the code that evaluates this is already separated from `GeometryEngine`). This is what I plan to do now, since it's easy. 

* Figure out some other way to propose atoms when the required torsions are absent. Longer term, but depending on special cases (probably in small molecule land), could be useful.

**Comment by jchodera** [comment]
Good catch!  

MCSS also lets you check for bonds in common---is there a way to express our requirement in terms of number of common bonds?  I'm thinking we need at least three atoms in a row in common, which may not be something that MCSS can check for.

Another possibility would be to expose a function in `GeometryEngine` that takes the molecule graphs and returns `True` or `False` (if there is no way to grow the atoms), or even returns a metric of number of atoms that have to be deleted or added.

**Comment by pgrinaway** [comment]
>Another possibility would be to expose a function in GeometryEngine that takes the molecule graphs and returns True or False (if there is no way to grow the atoms), or even returns a metric of number of atoms that have to be deleted or added.

We could just use the class that generates the proposal order to determine this.

**Comment by jchodera** [comment]
Good idea.

**Comment by pgrinaway** [comment]
Duplicate/resolved in #186
